### ISSUE TITLE
flake8 config error

### ISSUE THREAD

**Issue by ivirshup** [scverse/anndata#540]
There is a space in front of the `per-file-ignore` entry in the `flake8` config (added in https://github.com/theislab/scanpy/pull/1689). This is the same in scanpy, as the config here was just copied from there. This causes a parse failure which leads to the the `per-file-ignore`s becoming global instead. To demonstrate:

<details>
<summary> Parse failure </summary>

From the root of the repo

```python
>>> from flake8.api.legacy import get_style_guide
>>> get_style_guide().options.ignore
['#',
 'module',
 'imported',
 'but',
 'unused',
 '->',
 'required',
 'for',
 'Scanpys',
 'API',
 'F401',
 'W503',
 'W504',
 'E501',
 'E203',
 'E231',
 'E402',
 'E126',
 'E262',
 'E266',
 'E731',
 'E741',
 'per-file-ignores',
 '=',
 'tests/test*.py:',
 'F811',
 'anndata/compat/_overloaded_dict.py:',
 'F821',
 'anndata/tests/test_readwrite.py:',
 'E721']
```

</details>

Fixing the parse error reveals a number of flake8 errors which were accidentally being ignored. I came across this while adding back `F401`, since it should only be ignored for files which define a namespace.

It is kind of ironic that a tool for enforcement of code style is permissive to a fault about config parsing.

(ping @Zethson)

**Comment by flying-sheep** [comment]
@asottile can you help making such bugs impossible in the future? Implementing PyCQA/flake8#234 might be an easy way to do it.

**Comment by Zethson** [comment]
Nice catch @ivirshup. Sorry for this. I'll fix the config file.

**Comment by ivirshup** [comment]
@flying-sheep has the first issue here been resolved https://github.com/PyCQA/flake8/issues/234#issuecomment-812800722? If not, I don't think this is going to happen anytime soon. Even after this is implemented for `flake8`, other tools which read the `flake8` config would need to recognize the `pyproject.toml` entries (e.g. vscode, autopep8, etc.).

@Zethson no worries, all three of us managed to miss this. I'd really expect this sort of thing to cause a parse error.

**Comment by asottile** [comment]
really, this was all just a ploy to @ mention me about the pyproject.toml issue?  welcome to my blocklist

**Comment by flying-sheep** [comment]
@ivirshup yes it has been. pyproject.toml needs to have the `build-system` table to change pip’s behavior.

@asottile No, but I’m sorry I touched an apparently very emotional issue for you. I mentioned it because it’s the easy way. Manual validation would be another option that’s harder to implement. But if you really blocked me you’re not going to read this anyway …

**Comment by asottile** [comment]
> yes it has been. pyproject.toml needs to have the build-system table to change pip’s behavior.

it has not, the presence of the file alone still changes pip's behaviour in 21.0.1

**Comment by flying-sheep** [comment]
Oh interesting! I thought this was about when the presence of `pyproject.toml` made pip assume it’s not a legacy build (and therefore made it not use `setup.py`). They fixed that one long ago: Unless the `build-system` table is there, pip just assumes a `setup.py` build.

And I thought that’s that, but there seems to be something else at play. Do you have a link to the bug report about the behavior change you mean? Or can you describe it?

**Comment by flying-sheep** [comment]
Fixed in #863 and then superseded by #891
